### PR TITLE
Added abstract backend base with Gitlab implementation

### LINKED ISSUE #95: Create a proper BaseRepo backend which is API agnostic. Make Gitlab an implementation.
Edit:
In `baserepo.py`, there are three primary classes: `BaseRepo`, `Repo`, and `StudentRepo`. Those three classes form a backend for interfacing with Gitlab.

Let's make `BaseRepo` an abstract base of `Repo` and house it in `backends/base.py`. Additionally, let's add `BaseStudentRepo` which is a `BaseRepo` that's an abstract base for `StudentRepo`. We can also define the `BaseBackend` in this file which is an abstract base for a backend. `BaseBackend`, for now, will contain generically named class variables (such as `student_repo` and `repo`) so that assigner commands can be agnostic of the backend. Later this backend can do more fancy stuff.

Our current `Repo` and `StudentRepo` can then be implementations of `BaseRepo` and `BaseStudentRepo` respectively. They can be housed in `backends/gitlab.py`. A new `GitlabBackend` (`BaseBackend`) will also need to be created.

### PR DESCRIPTION
**Pull request by brhoades**
See description in #95 (fixes #95). I've added and separated backends as described there. I want to add a mock backend so I can test this locally next.

### CODE REVIEW

**Comment by brhoades** [comment]
I'll probably need ideas on how to handle Access. I have a feeling it should be moved to the backend.

File: assigner/backends/base.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    """A repo from which StudentRepos are cloned from (Homework Repo)."""
+    @classmethod
+    def new(cls, name: str, namespace: str, url_base: str, token: str):
+        raise NotImplementedError
+
+    def push_to(self, student_repo: StudentRepoBase, branch: str = "master") -> None:
+        raise NotImplementedError
+
+
+#pylint: disable=too-few-public-methods
```
Comment (LinuxMercedes) [inline assigner/backends/base.py]: IMO we can just add this one to the pylintrc -- that's an overly nitpicky lint message.

File: assigner/config.py [hunk 2]
```diff
@@ -4,9 +4,11 @@
 
 from collections import UserDict
 
+
 class DuplicateUserError(Exception):
     pass
 
+
```
Comment (brhoades) [inline assigner/config.py]: @LinuxMercedes Ideas about why this file wasn't linted?

File: assigner/backends/decorators.py [hunk 3]
```diff
@@ -0,0 +1,13 @@
+from assigner.backends import GitlabBackend
+from assigner.config import config_context
+
+
+@config_context
+def require_backend(config, func):
```
Comment (brhoades) [inline assigner/backends/decorators.py]: I'm not pleased with this line decorator. Not only does it not work right now (:D), it requires config which causes a file read. Then most of these commands require config too before this was called. I can see us opening and closing config 3 or 4 times for a single function call.

I'm definitely open to better ideas. I really like the decorator idea... maybe we should just cache a readonly config on the decorator object?

cc @LinuxMercedes

File: assigner/backends/base.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    @classmethod
+    def build_name(cls, semester: str, section: str, assignment: str, user: str) -> str:
+        raise NotImplementedError
+
+    def push(self, base_repo, branch: str = "master") -> None:
+        """Push base_repo code to this repo"""
+        raise NotImplementedError
+
+
+class TemplateRepoBase(RepoBase):
```
Comment (brhoades) [inline assigner/backends/base.py]: Open for better names. `BaseRepoBase` didn't sound good. `HomeworkRepoBase` didn't clear things up for me either.

File: assigner/backends/decorators.py (hunk 3 above)
Comment (brhoades) [inline assigner/backends/decorators.py]: @LinuxMercedes nerd snipe: write a type annotation for this decorator.

File: assigner/backends/base.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    def __init__(self, url_base: str, namespace: str, name: str, token: str, url: Optional[str] = None):
+        raise NotImplementedError
+
+    @classmethod
+    def build_url(cls, url_base: str, namespace: str, name: str) -> str:
+        """ Builds a url for a repository """
+        raise NotImplementedError
+
+    @classmethod
+    def from_url(cls, url: str, token: str):
```
Comment (brhoades) [inline assigner/backends/base.py]: I looked into returning a self type from a class method in type annotations. 

I think this is the recommended way:
https://github.com/python/mypy/issues/1212#issuecomment-250526301

Though I also saw this:
https://github.com/python/mypy/issues/1212#issuecomment-219253128

Thoughts? I really like these type annotations; I'll keep adding them where I can.

Comment (LinuxMercedes) [reply assigner/backends/base.py]: I think https://github.com/python/peps/commit/ada7d3566e26edf5381d1339b61e48a82c51c566 is what you want?

I also am a big fan of the type annotations; eventually I want to be able to integrate mypy here but there's a bit of footwork involved in writing annotations for some of the libraries we depend on.

File: assigner/backends/base.py (hunk 4 above)
Comment (LinuxMercedes) [reply assigner/backends/base.py]: I like `TemplateRepo`! We should make a note to update the wording in the documentation when we change this.

File: assigner/backends/base.py (hunk 5 above)
Comment (brhoades) [reply assigner/backends/base.py]: Yeah, I think that's what I want. I tried reading this at 7 last night and I couldn't comprehend it.

File: assigner/commands/init.py [hunk 6]
```diff
@@ -39,6 +39,7 @@ def guess_semester():
 
 @config_context
 def init(conf, _):
+    conf['backend'] = prompt("Backend to use (gitlab)", "gitlab")
```
Comment (LinuxMercedes) [inline assigner/commands/init.py]: Let's leave this choice out right now -- we'd have to rearchitect `init` anyway if we introduced a new backend since the next couple of config lines are GitLab-specific.

File: assigner/commands/lock.py [hunk 7]
```diff
@@ -15,6 +15,7 @@ def lock(args):
     #pylint: disable=no-value-for-parameter
     return manage_repos(
         args,
+        # TODO: Fix access
```
Comment (LinuxMercedes) [inline assigner/commands/lock.py]: Would the 'right' fix here be to have `RepoBase` (or maybe `StudentRepoBase`) implement `.lock()` and `.unlock()` methods and have the backend implementation handle the detail of what 'locking' and 'unlocking' means?

File: assigner/config.py (hunk 2 above)
Comment (LinuxMercedes) [reply assigner/config.py]: It passes lint here -- maybe we're missing the rule for this problem?

File: assigner/commands/lock.py (hunk 7 above)
Comment (brhoades) [reply assigner/commands/lock.py]: I think so. Since GitHub would have a different access concept and API. I really think that  the backend should decide what access permission specific things need (or what unlocking / locking mean, if anything).

File: assigner/config.py (hunk 2 above)
Comment (brhoades) [reply assigner/config.py]: Maybe... I keep getting this feeling that pyflakes or pylint aren't getting some files due to how they scan via modules. Flake8 throws an absolute fit on this project.

Comment (brhoades) [reply assigner/config.py]: Oh, unrelated, we should switch to flake8 :/. Doing some reading, pyflakes was meant to be embedded in some other project.

File: assigner/backends/base.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
+import git
+import re
+from typing import Optional
+
+
+class RepoError(Exception):
```
Comment (LinuxMercedes) [inline assigner/backends/base.py]: This doesn't have to be done in this PR but I would really like to do some better exception modeling for both GitLab and git--I'd really like to not have command implementations catch and interpret HTTPErrors or GitCommandErrors. Plus, I think that'd make testing easier--we can know what sorts of exceptions a function can throw* and test that Assigner handles them properly.

* except for abstract base classes...I dream of strongly typed algebraic effects but we don't even have those in Haskell yet 🙃

File: assigner/config.py (hunk 2 above)
Comment (LinuxMercedes) [reply assigner/config.py]: I threw an obvious lint problem in `config.py` and ran `pylint assigner` and it warned me about my problem, so I think we're good?

File: assigner/backends/base.py (hunk 1 above)
Comment (brhoades) [reply assigner/backends/base.py]: Done.

File: assigner/backends/base.py (hunk 4 above)
Comment (brhoades) [reply assigner/backends/base.py]: Changed. I updated documentation where I could find it.

File: assigner/config.py (hunk 2 above)
Comment (brhoades) [reply assigner/config.py]: Yeah :|

File: assigner/backends/decorators.py [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+def require_backend(func):
+    """Provides a backend depending on configuration."""
+    @config_context
+    def wrapper(config, cmdargs, *args, **kwargs):
+        if config.backend == "gitlab":
+            return func(config, GitlabBackend, cmdargs, *args, **kwargs)
+        if config.backend == "mock":
+            return func(config, MockBackend, cmdargs, *args, **kwargs)
+
```
Comment (LinuxMercedes) [inline assigner/backends/decorators.py]: Let's warn here so that users can fix their config.

File: assigner/backends/decorators.py [hunk 10]
```diff
@@ -0,0 +1,15 @@
+from assigner.backends import GitlabBackend, MockBackend
+from assigner.config import config_context
+
+
+def require_backend(func):
```
Comment (LinuxMercedes) [inline assigner/backends/decorators.py]: How about naming this `with_backend_and_config` and renaming `config_context` to `with_config`?

File: CHANGELOG.md [hunk 11]
```diff
@@ -9,6 +9,10 @@
 - Added Travis CI config for pylint and pyflakes.
 - Added unittest scaffolding with nose
 - Added `AssignerTestCase`, tests for `assigner`, `assigner get`, and all `--help` uses.
+- Added BackendBase, RepoBase, StudentRepoBase, and TemplateRepoBase.
+- Separated Gitlab backend from most of the code. Added Base implementations for Gitlab.
+- Added require_backend decorator with backend config option to load the desired backend.
+- pyflakes 1.6.0 upgrade for type annotations fixes.
```
Comment (LinuxMercedes) [inline CHANGELOG.md]: Let's add a note that "Base Repos" are now known as "Template Repos".

File: assigner/config.py [hunk 12]
```diff
@@ -22,6 +24,12 @@ class Config(UserDict):
 
         "type": "object",
         "properties": {
+            # Backend type (gitlab / mock)
+            "backend": {
+                "type": "string",
+                "pattern": "^(mock|gitlab)?$"
+            },
+
```
Comment (LinuxMercedes) [inline assigner/config.py]: Oh, we need to add some code to automatically migrate existing configs to new ones. I get the sense it's about time that we started versioning config files. That'd probably be good to do as a separate PR first?

File: assigner/backends/base.py (hunk 8 above)
Comment (LinuxMercedes) [reply assigner/backends/base.py]: This is issue #8 😄

File: assigner/backends/decorators.py (hunk 10 above)
Comment (brhoades) [reply assigner/backends/decorators.py]: I'm afraid of using 'with' in decorators, it sounds context managery. Maybe `requires_config_and_backend` and `requires_config` (I put an s in there too).

---
I spent more time looking at this than I should have. Most of the examples of decorators I look at are single words--- `patch`, `wraps`, `partial`, `classmethod`, `staticmethod`... I couldn't find more. I'm not sure if there's a wrong way to name them.

Comment (LinuxMercedes) [reply assigner/backends/decorators.py]: SGTM!

File: assigner/config.py (hunk 12 above)
Comment (LinuxMercedes) [reply assigner/config.py]: I'll take this on unless you really want to -- I have a few things I'd like to improve about the current state of config handling in general.

Comment (brhoades) [reply assigner/config.py]: Please, go ahead. I'm moving the next few days so I'll be not here.

File: CHANGELOG.md (hunk 11 above)
Comment (brhoades) [reply CHANGELOG.md]: Done.

File: assigner/backends/decorators.py (hunk 10 above)
Comment (brhoades) [reply assigner/backends/decorators.py]: Renamed as described.

File: assigner/backends/decorators.py (hunk 9 above)
Comment (brhoades) [reply assigner/backends/decorators.py]: I will add a check for no backend option. However, Config enforces that backend must match the regex:

```
WARNING  Your configuration is not valid: 'mocks' does not match '^(mock|gitlab)?$'
```

File: assigner/backends/base.py (hunk 1 above)
Comment (brhoades) [reply assigner/backends/base.py]: Yup, done.

File: assigner/backends/decorators.py [hunk 13]
```diff
@@ -1,15 +1,25 @@
...
 from assigner.backends import GitlabBackend, MockBackend
-from assigner.config import config_context
+from assigner.config import requires_config
+
+
+logger = logging.getLogger(__name__)
 
 
-def require_backend(func):
+def requires_backend_and_config(func):
```
Comment (LinuxMercedes) [inline assigner/backends/decorators.py]: NP 9000: Maybe name this `requires_config_and_backend` since it passes `config` first and `backend` second to the functions it wraps? Or we could change the order in which the args are passed.

File: assigner/backends/decorators.py (hunk 9 above)
Comment (LinuxMercedes) [reply assigner/backends/decorators.py]: True. I will handle some of this in the config versioning PR which I'm going to do today.

File: assigner/backends/decorators.py (hunk 13 above)
Comment (brhoades) [reply assigner/backends/decorators.py]: I ran sed for you.

**Comment by brhoades** [comment]
@LinuxMercedes I rebased off master and squashed some commits. Please look with your wizard eyes.

File: CHANGELOG.md [hunk 14]
```diff
@@ -14,6 +14,13 @@
 - Added Travis CI config for pylint and pyflakes.
 - Added unittest scaffolding with nose
 - Added `AssignerTestCase`, tests for `assigner`, `assigner get`, and all `--help` uses.
+- Added `BackendBase`, `RepoBase`, `StudentRepoBase`, and `TemplateRepoBase`.
```
Comment (LinuxMercedes) [inline CHANGELOG.md]: Can you move these lines up under the `Devel` header?

**Comment by LinuxMercedes** [comment]
Let me get #111 into shape and then we can put the last bit of polish on this!

Thank you so much for being patient and humoring my nitpicking ❤️

**Comment by brhoades** [comment]
I am the nitpick king, this is all fine.

File: CHANGELOG.md (hunk 14 above)
Comment (brhoades) [reply CHANGELOG.md]: I will after you merge the stuff that goes before this :D. Is it merged?

Comment (LinuxMercedes) [reply CHANGELOG.md]: This is waiting on #111 which is waiting for me to do the 1.2.0 release (#122) first 😃

**Comment by brhoades** [comment]
@LinuxMercedes changes made and rebased.

**Comment by LinuxMercedes** [comment]
I've merged #111! Let's rebase and I'll have a look-through and maybe use this a little bit before merging it.

**Comment by brhoades** [comment]
@LinuxMercedes 💢  😫  🔫 
```
Unmerged paths:
  (use "git reset HEAD <file>..." to unstage)
  (use "git add/rm <file>..." as appropriate to mark resolution)

        both modified:   assigner/__init__.py
        both modified:   assigner/backends/gitlab.py
        both modified:   assigner/commands/assign.py
        both modified:   assigner/commands/get.py
        both modified:   assigner/commands/init.py
        both modified:   assigner/commands/new.py
        both modified:   assigner/commands/open.py
        both modified:   assigner/commands/status.py
        deleted by us:   assigner/config.py
        both modified:   assigner/roster_util.py
```

File: assigner/backends/decorators.py [hunk 15]
```diff
@@ -0,0 +1,25 @@
+import logging
+
+from assigner.backends import GitlabBackend, MockBackend
+from assigner.config import requires_config
+
+
+logger = logging.getLogger(__name__)
+
+
+def requires_config_and_backend(func):
```
Comment (LinuxMercedes) [inline assigner/backends/decorators.py]: I don't like having this and `requires_config` in different places, but I'm not sure what the right place is for them.

File: assigner/backends/decorators.py [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+    """Provides a backend depending on configuration."""
+    @requires_config
+    def wrapper(config, cmdargs, *args, **kwargs):
+        try:
+            config.backend
+        except KeyError:
```
Comment (LinuxMercedes) [inline assigner/backends/decorators.py]: I think I'd rather propagate this `KeyError` out; if they haven't set a backend, none of this is going to work...

File: assigner/backends/mock.py [hunk 17]
```diff
@@ -0,0 +1,270 @@
...
+        return "branch"
+
+    def archive(self):
+        return MagicMock()
+
+    def unarchive(self):
+        return MagicMock()
+
+    def unlock(self, student_id: str) -> None:
+        self.edit_member(student_id, Access.reporter)
```
Comment (LinuxMercedes) [inline assigner/backends/mock.py]: Should this set access to `developer` (and `lock` below switch to `reporter`)?

File: assigner/backends/decorators.py (hunk 15 above)
Comment (LinuxMercedes) [reply assigner/backends/decorators.py]: Ehh this is a perfectly fine color for this yak shed.

File: assigner/commands/lock.py [hunk 18]
```diff
@@ -15,7 +14,7 @@ def lock(args):
     #pylint: disable=no-value-for-parameter
     return manage_repos(
         args,
-        lambda repo, student: repo.edit_member(student["id"], Access.reporter)
+        lambda repo, student: repo.unlock(student["id"])
```
[the remark this answers is not in the document]
Comment (brhoades) [reply assigner/commands/lock.py]: Yeah. I think this was a typo and should be what unlock is.

**Comment by LinuxMercedes** [comment]
@brhoades The only thing I think I have left to do for this is to finish up the last commit -- adding more return type signatures. Care to give it a once-over and an LGTM? :heart:

**Comment by brhoades** [comment]
LGTM :+1:

**Comment by LinuxMercedes** [comment]
:tada: :tada: :tada: :tada: :tada:
